### ISSUE TITLE
[Breaking change]: C# overload resolution prefers `params` *span type* overloads which cannot be used in Expression lambdas

### ISSUE THREAD

**Issue by cston** [dotnet/docs#43949]
### Description

C# 13 adds support for `params` parameters declared with collection types other than arrays. In particular, `params ReadOnlySpan<T>` and `params Span<T>` are supported, and overload resolution will prefer a `params` *span type* over a `params` *array type* when both are applicable.

.NET 9 adds `params` span overloads for various methods in the BCL, in addition to the existing overloads with `params` arrays. When recompiling existing callers of those methods, where arguments are passed in *expanded form*, the compiler will now bind to the `params` span overload.

That leads to a potential breaking change for existing calls to those overloads within `Expression` lambda expressions, where `ref struct` instances are not supported. In those cases, the C# 13 compiler will report an error when binding to the `params` span overload.

For instance, consider `string.Join()` (see https://github.com/dotnet/runtime/issues/110592):
```csharp
using System;
using System.Linq.Expressions;

Expression<Func<string, string, string>> join = (x, y) => string.Join("", x, y);
```

When compiled with .NET 8, the call binds to `String.Join(string? separator, params string?[] value)`, without errors.

When compiled with C# 13 and .NET 9, the call binds to `String.Join(string? separator, params ReadOnlySpan<string?> value)`, and because the call is within an `Expression` tree, the following errors are reported:
```
error CS8640: Expression tree cannot contain value of ref struct or restricted type 'ReadOnlySpan'.
error CS9226: An expression tree may not contain an expanded form of non-array params collection parameter.
```

### Version

.NET 9

### Previous behavior

Previously, `params` overloads were limited to *array types* only, so calls to those methods in *expanded form* resulted in implicit arrays.

### New behavior

With C#13 and .NET 9, for methods with overloads that take `params` array and `params` span, overload resolution will prefer the `params` span overload, resulting in an implicit span instance. For calls within `Expression` lambda expressions, the implicit `ref struct` span instance will be reported as an error.

### Type of breaking change

- [ ] **Binary incompatible**: Existing binaries might encounter a breaking change in behavior, such as failure to load or execute, and if so, require recompilation.
- [x] **Source incompatible**: When recompiled using the new SDK or component or to target the new runtime, existing source code might require source changes to compile successfully.
- [ ] **Behavioral change**: Existing binaries might behave differently at run time.

### Reason for change

 `params` span support allows the compiler to avoid unnecessary allocation for the `params` argument.

### Recommended action

The recommended work around is to call the method with an explicit array, so the call is bound to the `params` array overload.

For the example above, use `new string[] { ... }`:
```csharp
Expression<Func<string, string, string>> join = (x, y) => string.Join("", new string[] { x, y });
```

### Feature area

Other (please put exact area in description textbox)

### Affected APIs

```csharp
String.Join(string? separator, params string?[] value)
// ...
```

**Comment by gewarren** [comment]
@cston Should the syntax block for this method say `params` instead of `scoped` for the second parameter?

https://learn.microsoft.com/en-us/dotnet/api/system.string.join?view=net-9.0#system-string-join(system-string-system-readonlyspan((system-object)))

**Comment by cston** [comment]
> (https://github.com/cston) Should the syntax block for this method say `params` instead of `scoped` for the second parameter?

Yes, it looks like that method should be `params` rather than `scoped`.

**Comment by gewarren** [comment]
@BillWagner Is 👆 a known issue?

**Comment by BillWagner** [comment]
@gewarren 

This change is noted here: https://github.com/dotnet/roslyn/blob/main/docs/compilers/CSharp/Compiler%20Breaking%20Changes%20-%20DotNet%2010.md#spant-and-readonlyspant-overloads-are-applicable-in-more-scenarios-in-c-14-and-newer

I plan to get that published in early January.  It will be a node similar to [here](https://learn.microsoft.com/en-us/dotnet/csharp/whats-new/breaking-changes/compiler%20breaking%20changes%20-%20dotnet%209), which has the breaking changes since .NET 8 GA.

**Comment by gewarren** [comment]
@BillWagner Sorry, I was asking about the `scoped` vs `params` issue. This is what the IL looks like, as opposed to the params array methods which actually have `params` in the IL. So I guess that's why the docs aren't reading/showing it correctly:

```
public static string Concat([ParamCollection][ScopedRef] ReadOnlySpan<object?> args)
{
	throw null;
}
```

**Comment by cston** [comment]
> This change is noted here: https://github.com/dotnet/roslyn/blob/main/docs/compilers/CSharp/Compiler%20Breaking%20Changes%20-%20DotNet%2010.md#spant-and-readonlyspant-overloads-are-applicable-in-more-scenarios-in-c-14-and-newer

@BillWagner, the breaking change here is specific to the *`params` span* feature in C#13 rather than *first-class spans* and, more specifically, for calls to methods with `params` span overloads within `Expression` trees.

**Comment by gewarren** [comment]
I'm not seeing the TagList.Add() method in the docs: https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.taglist.add?view=net-9.0#system-diagnostics-taglist-add(system-collections-generic-keyvaluepair((system-string-system-object)))

**Comment by cston** [comment]
> Some of these didn't have pre-existing `params` array overloads, so they wouldn't be affected by the breaking change, correct? If so, I'll leave those out of the article. For example, the CollectionExtensions methods.

No, if there was no corresponding array overload, then the method should not be affected by the breaking change.

> I'm not seeing the TagList.Add() method in the docs: ...

That was a copy-paste error. I've removed it from list above.

Thanks for catching those cases.
